lib/rules/disallow-newline-before-block-statements.js

```diff
@@ -90,15 +184,82 @@ module.exports.prototype = {
     },
 
     check: function(file, errors) {
-        file.iterateNodesByType('BlockStatement', function(node) {
+        if (this._setting === true) {
+            checkAll(file, errors);
+        } else {
+            checkSpecificTypes(file, errors, this._setting);
+        }
+    }
+};
+
+function checkAll(file, errors) {
+    file.iterateNodesByType('BlockStatement', function(node) {
+        var openingBrace = file.getFirstNodeToken(node);
+        var prevToken = file.getPrevToken(openingBrace);
+
+        errors.assert.sameLine({
+            token: prevToken,
+            nextToken: openingBrace,
+            message: 'Newline before curly brace for block statement is disallowed'
+        });
+    });
+}
+
+function checkSpecificTypes(file, errors, types) {
+    file.iterateNodesByType(['BlockStatement'], function(node) {
+        var type = getBlockType(node);
+        if (types.indexOf(type) !== -1) {
             var openingBrace = file.getFirstNodeToken(node);
             var prevToken = file.getPrevToken(openingBrace);
 
             errors.assert.sameLine({
                 token: prevToken,
                 nextToken: openingBrace,
-                message: 'Newline before curly brace for block statement is disallowed'
+                message: 'Newline before curly brace for "' + type + '" block statement is disallowed'
             });
-        });
+        }
+    });
+}
+
+function getBlockType(node) {
+    var parentType = node.parentNode.type;
+    var blockType;
+    switch (parentType) {
+        case 'IfStatement':
+            if (node.parentNode.alternate === node) {
+                blockType = 'else';
+            } else {
+                blockType = 'if';
+            }
+            break;
+        case 'FunctionDeclaration':
+        case 'FunctionExpression':
+            blockType = 'function';
+            break;
+        case 'ForStatement':
+        case 'ForInStatement':
+        case 'ForOfStatement':
+            blockType = 'for';
+            break;
+        case 'WhileStatement':
+            blockType = 'while';
+            break;
+        case 'DoWhileStatement':
+            blockType = 'do';
+            break;
+        case 'TryStatement':
+            if (node.parentNode.finalizer === node) {
```

Comment:
this whole block is: `return node.parentNode.finalizer === node ? 'finally' : 'try'`